File: common/scala/src/main/scala/org/aiddl/common/scala/planning/state_variable/ApplicableFunction.scala
package org.aiddl.common.scala.planning.state_variable

import scala.collection.mutable.Set
import scala.collection.mutable.Map
import scala.collection.mutable.HashMap
import scala.collection.mutable.HashSet

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.function.Initializable
import org.aiddl.core.scala.function.Configurable

import org.aiddl.core.scala.container.Container

import org.aiddl.core.scala.representation._

import org.aiddl.common.scala.planning.PlanningTerm._

import Term.given_Conversion_Term_KeyVal
import scala.language.implicitConversions

class ApplicableFunction extends Function {
  def apply( x: Term ): Term = x match {
    case Tuple( a: Tuple, s: SetTerm ) => this(a, s)
    case _ => ???
  }

  def apply( a: Tuple, s: SetTerm ): Term = {
    Bool(a(Preconditions).asCol.forall( sva => s.get(sva.key) match {
      case Some(v) => v == sva.value
      case None => false
    }))
  }
}

File: common/java/src/main/java/org/aiddl/common/java/planning/state_variable/ApplicableFunction.java
package org.aiddl.common.java.planning.state_variable;

import org.aiddl.common.java.planning.PlanningTerm;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.interfaces.InterfaceImplementation;
import org.aiddl.core.java.representation.SetTerm;
import org.aiddl.core.java.representation.SymbolicTerm;
import org.aiddl.core.java.representation.Term;

public class ApplicableFunction implements Function, InterfaceImplementation {
	private final static SymbolicTerm InterfaceUri = Term.sym("org.aiddl.common.planning.state-variable.applicable");
	
	@Override
	public SymbolicTerm getInterfaceUri() {
		return InterfaceUri;
	}
	
	@Override
	public Term apply(Term args) {
		Term a = args.get(0);
		SetTerm s = args.get(1).asSet();

		for ( Term p : a.get(PlanningTerm.Preconditions).asCollection() ) {
			Term p_sv = p.getKey();
			Term p_a = p.getValue();
			Term s_a = s.get(p_sv);
			if ( s_a == null || !s_a.equals(p_a) ) {
				return Term.bool(false);
			}
		}
		return Term.bool(true);
	}
}
